### PR TITLE
Smoke:Enable TCBRC in ffmpeg-qsv upstream

### PR DESCRIPTION
**Pull request by Wanghangjie8**
Smoke:Enable TCBRC in ffmpeg-qsv upstream

Signed-off-by: Wang Hangjie <hangjiex.wang@intel.com>

### CODE REVIEW

File: lib/ffmpeg/encoderbase.py [hunk 1]
```diff
@@ -57,6 +57,7 @@ def hwformat(self):
   level         = property(lambda s: s.ifprop("level", " -level {level}"))
   ladepth       = property(lambda s: s.ifprop("ladepth", " -look_ahead 1 -look_ahead_depth {ladepth}"))
   forced_idr    = property(lambda s: s.ifprop("vforced_idr", " -forced_idr 1 -force_key_frames expr:1"))
+  tcbrc_idr     = property(lambda s: s.ifprop("vtcbrc_idr", " -low_delay_brc 1 -low_power 1"))
```
Comment (FocusLuo) [inline lib/ffmpeg/encoderbase.py:60]: tcbrc is a BRC mode, its name should not include idr which will mistake others

File: test/ffmpeg-qsv/encode/avc.py [hunk 2]
```diff
@@ -241,6 +241,26 @@ def test(self, case, rcmode, bitrate, maxrate, qp, quality, profile):
...
+  def init(self, tspec, case, rcmode, bitrate, maxrate, qp, quality, profile):
+    vars(self).update(tspec[case].copy())
+    vars(self).update(
+      rcmode     = rcmode,
+      bitrate    = bitrate,
+      case       = case,
+      maxrate    = maxrate,
+      minrate    = bitrate,
+      profile    = profile,
+      qp         = qp,
```
Comment (FocusLuo) [inline test/ffmpeg-qsv/encode/avc.py:254]: tcbrc is a BRC, currently it is low power BRC. it should not include QP.

**Comment by FocusLuo** [comment]
Based on your patch, It is better to update your patch tittle to remove "smoke", and to include some key info such "AVC TCBRC" support

**Comment by FocusLuo** [comment]
You should add some description for your patch's commit, for example what this patch enable/fix something

File: lib/parameters.py [hunk 3]
```diff
@@ -213,6 +213,34 @@ def gen_avc_forced_idr_parameters(spec, profiles):
   params = gen_avc_forced_idr_variants(spec, profiles)
   return keys, params
 
+def gen_avc_tcbrc_idr_variants(spec, profiles):
```
Comment (FocusLuo) [inline lib/parameters.py:216]: TCBRC is a BRC mode, you should refer VBR, but not CQP to generate test params

**Comment by uartie** [comment]
@FocusLuo are there patches in ffmpeg that handle TCBRC?

**Comment by uartie** [comment]
> @FocusLuo are there patches in ffmpeg that handle TCBRC?

`grep -R TCBRC ffmpeg/.` returns nothing

**Comment by FocusLuo** [comment]
@uartie maybe need to search "low_delay_brc"
It uses "VBR + "-low_delay_brc 1 -low_power 1"" in ffmpeg to control BRC and package size. 

TCBRC: it is used for limited bandwidth. By limiting packet size, the packet can be send/received with low delay. For example if you set "-r 30 -maxrate 7500000 -low_delay_brc 1", the average frame size is limited under 7500000/8*30 = 31250.

e.g.
./ffmpeg -v verbose -f rawvideo -pix_fmt yuv420p -s:v 1280x720 -r 30 -i test.yuv -c:v h264_qsv -b:v 5000000 -maxrate 7500000 -low_delay_brc 1 -low_power 1 -y output.h264

**Comment by FocusLuo** [comment]
> > @FocusLuo are there patches in ffmpeg that handle TCBRC?
> 
> `grep -R TCBRC ffmpeg/.` returns nothing

@uartie  How about use "grep -r tcbrc test/ffmpeg-qsv/."

File: lib/ffmpeg/encoderbase.py (hunk 1 above)
Comment (Wanghangjie8) [reply lib/ffmpeg/encoderbase.py]: Ok, I have revised and resubmitted

File: test/ffmpeg-qsv/encode/avc.py (hunk 2 above)
Comment (Wanghangjie8) [reply test/ffmpeg-qsv/encode/avc.py]: Ok, I have revised and resubmitted

File: lib/parameters.py (hunk 3 above)
Comment (Wanghangjie8) [reply lib/parameters.py]: Ok, I have revised and resubmitted

**Comment by uartie** [comment]
> > > @FocusLuo are there patches in ffmpeg that handle TCBRC?
> > 
> > 
> > `grep -R TCBRC ffmpeg/.` returns nothing
> 
> @uartie How about use "grep -r tcbrc test/ffmpeg-qsv/."

No, I meant it in the context of the ffmpeg sourcecode tree... not vaapi-fits.

File: lib/ffmpeg/encoderbase.py [hunk 4]
```diff
@@ -57,6 +57,7 @@ def hwformat(self):
   level         = property(lambda s: s.ifprop("level", " -level {level}"))
   ladepth       = property(lambda s: s.ifprop("ladepth", " -look_ahead 1 -look_ahead_depth {ladepth}"))
   forced_idr    = property(lambda s: s.ifprop("vforced_idr", " -forced_idr 1 -force_key_frames expr:1"))
+  tcbrc         = property(lambda s: s.ifprop("vtcbrc", " -low_delay_brc 1 -low_power 1"))
```
Comment (uartie) [inline lib/ffmpeg/encoderbase.py:60]: tcbrc should not be a property, it should be an accepted value for rcmode.  Please treat it as such.

**Review by uartie — CHANGES_REQUESTED** [review]

File: lib/ffmpeg/encoderbase.py [hunk 5]
```diff
@@ -96,6 +97,16 @@ def encparams(self):
...
+  @property
+  def encparams(self):
+    return (
+      f"{self.profile}{self.qp}{self.quality}{self.gop}"
+      f"{self.bframes}{self.slices}{self.minrate}{self.maxrate}{self.refs}"
+      f"{self.extbrc}{self.loopshp}{self.looplvl}{self.tilecols}{self.tilerows}"
+      f"{self.level}{self.ladepth}{self.tcbrc}{self.intref}{self.lowpower}"
+      f"{self.maxframesize}{self.bstrategy}{self.pict}"
+    )
+
```
Comment (uartie) [inline lib/ffmpeg/encoderbase.py:109]: why redefined `encparams`?

File: lib/ffmpeg/encoderbase.py [hunk 6]
```diff
@@ -316,3 +329,108 @@ def check_max_frame_size(self):
     frameSizes = re.findall(r'(?<=pkt_size=).[0-9]*', output)
     for frameSize in frameSizes:
       assert (self.maxframesize * 1000) >= int(frameSize), "It appears that the max_frame_size did not work"
+ 
+ def encode(self):
```
Comment (uartie) [inline lib/ffmpeg/encoderbase.py:333]: Why redefined `encode`

**Review by uartie — CHANGES_REQUESTED** [review]
Several functions have been redefined... please don't.

**Review by uartie — CHANGES_REQUESTED** [review]

**Comment by Wanghangjie8** [comment]
Sorry, I found that my patch is mentioned in the wrong place, I have just mentioned the latest patch up.

**Comment by Wanghangjie8** [comment]
@uartie I have made the changes here. Do you have any other suggestions?

**Comment by Wanghangjie8** [comment]
@uartie  I have made the changes here. Do you have any other suggestions?

File: lib/ffmpeg/qsv/encoder.py [hunk 7]
```diff
@@ -17,6 +17,7 @@ class Encoder(FFEncoder):
   hwaccel = property(lambda s: "qsv")
   tilecols = property(lambda s: s.ifprop("tilecols", " -tile_cols {tilecols}"))
   tilerows = property(lambda s: s.ifprop("tilerows", " -tile_rows {tilerows}"))
+  tcbrc    = property(lambda s: s.ifprop("tcbrc", " -tc_brc {tcbrc}"))
```
Comment (uartie) [inline lib/ffmpeg/qsv/encoder.py:20]: Where is `-tc_brc` option defined in ffmpeg?  Is there a patch pending somewhere? mainline? cartwheel?

File: lib/parameters.py [hunk 8]
```diff
@@ -213,6 +213,31 @@ def gen_avc_forced_idr_parameters(spec, profiles):
...
+      cprofiles = [uprofile] if uprofile else profiles
+      rcmode  = variant["rcmode"]
+      bitrate = variant.get("bitrate", None)
+      if "cqp" == rcmode:
+        assert bitrate == None, "We shouldn't set a value to bitrate for CQP."
+        variant.update(maxrate = None)
+        if "cbr" == rcmode:
+          variant.update(maxrate = bitrate)
+        elif "vbr" == rcmode:
+          variant.update(maxrate = bitrate * 2)
```
Comment (uartie) [inline lib/parameters.py:229]: Now I'm confused... previous discussions state that tcbrc is only valid for VBR low power mode with ldb.

**Comment by uartie** [comment]
Please see #442 for example of https://github.com/intel/vaapi-fits/pull/431#discussion_r941994192

File: lib/ffmpeg/encoderbase.py (hunk 4 above)
Comment (uartie) [reply lib/ffmpeg/encoderbase.py]: See #442

**Comment by uartie** [comment]
Superseded by #442

**Pull request closed without merging**